doc/bitlbee.8
.\" BitlBee is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH bitlbee 8 "19 May 2010"
.SH NAME
BitlBee \- IRC gateway to IM chat networks
.SH SYNOPSIS
.PP
.B bitlbee
[\-I]
[\-c \fIconfiguration file\fP]
[\-d \fIconfiguration directory\fP]
.PP
.B bitlbee
\-D
[\-i \fIaddress\fP]
[\-p \fIport number\fP]
[\-n]
[\-v]
[\-c \fIconfiguration file\fP]
[\-d \fIconfiguration directory\fP]
.PP
.B bitlbee
\-h
.RI
.SH DESCRIPTION
BitlBee is an IRC daemon that can talk to instant messaging 
networks and acts as a gateway. Users can connect to the server
with any normal IRC client and see their 'buddy list' in
&bitlbee.

\fBbitlbee\fP is normally started as a system service (with init
scripts, systemd units or equivalent). It can also be run as an
.BR inetd (8)
/
.BR xinetd (8)
service or a stand-alone daemon with the options below.

To use it, connect to it with an IRC client with a command such as
"/connect localhost"

.PP
.SH RUN MODES
.PP
.IP "-I"
Run in 
.BR inetd (8)
mode. This is the default setting for historical reasons, but not the
recommended one (See ForkDaemon below).
.IP "-D"
Run in \fBDaemon\fP mode. In this mode, BitlBee forks to the background and
waits for new connections. All clients will be served from one process.
This mode is also useful for debugging.
.IP "-F"
Run in \fBForkDaemon\fP mode. This is similar to ordinary daemon mode, but every
client gets its own process. Easier to set up than inetd mode, and without
the possible stability issues. This is the recommended runmode for most
use cases.

.PP
.SH OPTIONS
.PP

.IP "-i \fIaddress\fP"
When running in daemon mode, specifies the network interface
(identified by IP address) to which the daemon should attach. Use this if
you don't want BitlBee to listen on every interface (which is the default
behaviour).
.IP "-p \fIport number\fP"
When running in daemon mode, specifies the port number on
which BitlBee should listen for connections. 6667 is the default value.
.IP "-n"
When running in daemon mode, this option prevents BitlBee from
forking into the background.
.IP "-v"
Be more verbose. This only works together with the \fB-n\fP flag.
.IP "-c \fIpath to other configuration file\fP"
Use a different configuration file.
.IP "-d \fIpath to user settings directory\fP"
BitlBee normally saves every user's settings in \fB/var/lib/bitlbee/\fP. If
you want the settings to be stored somewhere else (for example, if you don't
have write permissions in the default location), use this option.
.IP "-h"
Show help information.
.SH COMMANDS
To get a complete list of commands, please use the \fBhelp commands\fP
command in the &bitlbee channel after connecting to bitlbee.
.SH "SEE ALSO"
.BR inetd (8), 
.BR inetd.conf (5)

.BR https://www.bitlbee.org/

For more information on using BitlBee, once connected, you should use
the \fBhelp\fP command in the &bitlbee channel.
.SH BUGS
Of course there are bugs. If you find some, please report them at
\fBhttps://bugs.bitlbee.org/\fP.
.SH LICENSE
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful, but
\fBWITHOUT ANY WARRANTY\fR; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU 
General Public License for more details.
.PP
You should have received a copy of the GNU General Public License 
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple PLace, Suite 330, Boston, MA  02111-1307  USA
.SH AUTHORS
.PP
 Wilmer van der Gaast <wilmer@gaast.net>


skype/NEWS
VERSION         DESCRIPTION
-----------------------------------------------------------------------------
0.8.4         - now using python2.7 directly in case python would point to python3k
              - merge patch to avoid a crash when failing to connect to skyped
              - merge support for building the plugin on NetBSD
              - merge Debian patches
0.8.3         - support for BitlBee 1.3dev
              - fixed --debug switch (-d was fine)
              - documentation fixes
0.8.2         - building documentation is now optional
              - new settings: test_join and show_moods
              - '~' in skyped.conf is now expanded to the user's home directory
	      - groupchat channel names are now persistent (requires
	        BitlBee-1.2.6)
0.8.1         - support for BitlBee 1.2.5
              - support for Skype 2.1.0.81 and Skype4Py 1.0.32.0
              - the plugin part now supports FreeBSD
              - fix for edited messages, the prefix can now be configured
0.8.0         - fix build on x86_64 (-fPIC usage)
              - debug messages now have a timestamp
              - more work on having the default config under ~/.skyped
              - added a manual page for skyped
0.7.2         - add --log option to skyped to allow logging while it the
                daemon is in the background.
              - prefer config files from ~/.skyped over /etc/skyped
              - handle the case when LANG and LC_ALL env vars are empty
0.7.1         - mostly internal changes, the monster read callback is
                now replaced by tiny parser functions
0.7.0         - made 'make config' more portable
              - add 'skypeconsole' buddy for debugging purposes
              - support autojoin for bookmarked groupchats
              - skyped: make hardwired '/dev/null' portable and fix
                Python-2.6 warnings
0.6.3         - various osx-specific improvements (see the new screenshot!)
              - added python-gnutls install instructions
              - bitlbee.pc is now searched under
                /usr/local/lib/pkgconfig by default to help LFS monkeys ;)
0.6.2         - bugfix: make install required the plugin even in case
                its build was disabled
0.6.1         - added keepalive traffic to avoid disconnects in bitlbee
                when there is no traffic for a long time
              - now the plugin or skyped is automatically disabled if
		the dependencies are not available; useful in case the
		plugin is to be installed on a public server, or the
		skyped is to be used with a public server only
0.6.0         - works with BitlBee 1.2.1
0.5.1         - configure now automatically detects the right prefix to
                match witl BitlBee's one
              - minor documentation improvements (public chats, bug
                reporting address)
0.5.0         - skyped now uses gnutls if possible, which seem to be
                more stable, compared to openssl.
              - skyped now tries to handle all read/write errors from/to
                clients, and always just warn about failures, never exit.
              - installation for Debian users should be more simple
              - improved documentation
              - this .0 release should be quite stable, only about 100
                lines of new code
0.4.2         - skyped should be now more responsive
              - new skypeout_offline setting for hiding/showing SkypeOut
                contacts
              - support for SkypeOut calls
              - support for querying the balance from Skype
              - all setting should be documented now
0.4.1         - support for building the plugin on Mac OSX
              - tested with BitlBee 1.2 and Skype 2.0.0.63
              - avoid ${prefix} (by autoconf) in the config file as we
                don't handle such a variable
              - now you can call echo123 (patch by Riskó Gergely)
0.4.0         - support for starting, accepting and rejecting calls
              - also updated documentation (the key is the account set
                skype/call command)
	      - as usual with the .0 releases, be careful, ~200 lines of
	        new code
0.3.2         - support for Skype 2.0.0.43
              - skyped now automatically starts/shuts down skype
              - improved 'make prepare' to handle more automake versions
              - documentation improvements
0.3.1         - beautify output when skyped is interrupted using ^C
              - 'nick skype foo' now really sets display name, not the mood
                text
              - documentation fixups
              - this version should be again as stable as 0.2.6 was
0.3.0         - authentication support in skyped via ssl
              - ~200 lines of new code, so be careful :)
	      - upgraders: please read the documentation about how to set up
	        your config the ssl certificate, this was no necessary till now
0.2.6         - the server setting has a default value, 'localhost' so in most
		cases you no longer have to set it explicitly
	      - support for receiving emoted messages, ie. when the user types
	        '/me foo'
	      - support for setting the display name (nick 0 "foo bar") - it
	        sets the mood text
0.2.5         - now bitlbee's info command is supported (it displays full name,
		birthday, homepage, age, etc.)
0.2.4         - improve documentation based on feedback from people on #bitlbee
	      - fixed for Skype4Py >= 0.9.28.4
	      - tested with latest Skype beta, too (the one which supports
	        video)
0.2.3         - fixed that annoying "creating groupchat failed" warning
0.2.2         - don't change the topic if skype does not report a successful
                topic change
              - fixed for the recent bitlbee API changes
0.2.1         - topic support in group chats
              - bugfixes for multiline messages
              - this version should be again as stable as 0.1.4 was

bitlbee.c
/********************************************************************\
  * BitlBee -- An IRC to other IM-networks gateway                     *
  *                                                                    *
  * Copyright 2002-2004 Wilmer van der Gaast and others                *
  \********************************************************************/

/* Main file                                                            */

/*
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License with
  the Debian GNU/Linux distribution in /usr/share/common-licenses/GPL;
  if not, write to the Free Software Foundation, Inc., 51 Franklin St.,
  Fifth Floor, Boston, MA  02110-1301  USA
*/

#define BITLBEE_CORE
#include "bitlbee.h"
#include "commands.h"
#include "protocols/nogaim.h"
#include "help.h"
#include "ipc.h"
#include <signal.h>
#include <stdio.h>
#include <errno.h>

static gboolean bitlbee_io_new_client(gpointer data, gint fd, b_input_condition condition);

static gboolean try_listen(struct addrinfo *res)
{
	int i;

	global.listen_socket = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
	if (global.listen_socket < 0) {
		log_error("socket");
		return FALSE;
	}

#ifdef IPV6_V6ONLY
	if (res->ai_family == AF_INET6) {
		i = 0;
		setsockopt(global.listen_socket, IPPROTO_IPV6, IPV6_V6ONLY,
		           (char *) &i, sizeof(i));
	}
#endif

	/* TIME_WAIT (?) sucks.. */
	i = 1;
	setsockopt(global.listen_socket, SOL_SOCKET, SO_REUSEADDR, &i, sizeof(i));

	i = bind(global.listen_socket, res->ai_addr, res->ai_addrlen);
	if (i == -1) {
		closesocket(global.listen_socket);
		global.listen_socket = -1;

		log_error("bind");
		return FALSE;
	}

	return TRUE;
}

int bitlbee_daemon_init()
{
	struct addrinfo *res, hints, *addrinfo_bind;
	int i;
	FILE *fp;

	log_link(LOGLVL_ERROR, LOGOUTPUT_CONSOLE);
	log_link(LOGLVL_WARNING, LOGOUTPUT_CONSOLE);

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = PF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_flags = AI_PASSIVE
#ifdef AI_ADDRCONFIG
	                 /* Disabled as it may be doing more harm than good: this flag
	                    ignores IPv6 addresses on lo (which seems reasonable), but
	                    the result is that some clients (including irssi) try to
	                    connect to ::1 and fail.
	                                | AI_ADDRCONFIG */
#endif
	;

	i = getaddrinfo(global.conf->iface_in, global.conf->port, &hints, &addrinfo_bind);
	if (i) {
		log_message(LOGLVL_ERROR, "Couldn't parse address `%s': %s",
		            global.conf->iface_in, gai_strerror(i));
		return -1;
	}

	global.listen_socket = -1;

	/* Try IPv6 first (which will become an IPv6+IPv4 socket). */
	for (res = addrinfo_bind; res; res = res->ai_next) {
		if (res->ai_family == AF_INET6 && try_listen(res)) {
			break;
		}
	}

	/* The rest (so IPv4, I guess). */
	if (res == NULL) {
		for (res = addrinfo_bind; res; res = res->ai_next) {
			if (res->ai_family != AF_INET6 && try_listen(res)) {
				break;
			}
		}
	}

	freeaddrinfo(addrinfo_bind);

	if (global.listen_socket == -1 && errno == EADDRINUSE) {
		log_message(LOGLVL_ERROR, "Can't listen on port %s. Is another bitlbee already running?",
		            global.conf->port);
		return(-1);
	}

	i = listen(global.listen_socket, 10);
	if (i == -1) {
		log_error("listen");
		return(-1);
	}

	global.listen_watch_source_id = b_input_add(global.listen_socket, B_EV_IO_READ, bitlbee_io_new_client, NULL);

	if (!global.conf->nofork) {
		i = fork();
		if (i == -1) {
			log_error("fork");
			return(-1);
		} else if (i != 0) {
			exit(0);
		}

		setsid();
		i = chdir("/");
		/* Don't use i, just make gcc happy. :-/ */

		if (getenv("_BITLBEE_RESTART_STATE") == NULL) {
			for (i = 0; i < 3; i++) {
				if (close(i) == 0) {
					/* Keep something bogus on those fd's just in case. */
					open("/dev/null", O_WRONLY);
				}
			}
		}
	}

	if (global.conf->runmode == RUNMODE_FORKDAEMON) {
		ipc_master_load_state(getenv("_BITLBEE_RESTART_STATE"));
	}

	if (global.conf->runmode == RUNMODE_DAEMON || global.conf->runmode == RUNMODE_FORKDAEMON) {
		ipc_master_listen_socket();
	}

	if ((fp = fopen(global.conf->pidfile, "w"))) {
		fprintf(fp, "%d\n", (int) getpid());
		fclose(fp);
	} else {
		log_message(LOGLVL_WARNING, "Warning: Couldn't write PID to `%s'", global.conf->pidfile);
	}

	if (!global.conf->nofork) {
		log_link(LOGLVL_ERROR, LOGOUTPUT_SYSLOG);
		log_link(LOGLVL_WARNING, LOGOUTPUT_SYSLOG);
	}

	return(0);
}

int bitlbee_inetd_init()
{
	if (!irc_new(0)) {
		return(1);
	}

	return(0);
}

gboolean bitlbee_io_current_client_read(gpointer data, gint fd, b_input_condition cond)
{
	irc_t *irc = data;
	char line[513];
	int st;

	st = read(irc->fd, line, sizeof(line) - 1);
	if (st == 0) {
		irc_abort(irc, 1, "Connection reset by peer");
		return FALSE;
	} else if (st < 0) {
		if (sockerr_again()) {
			return TRUE;
		} else {
			irc_abort(irc, 1, "Read error: %s", strerror(errno));
			return FALSE;
		}
	}

	line[st] = '\0';
	if (irc->readbuffer == NULL) {
		irc->readbuffer = g_strdup(line);
	} else {
		irc->readbuffer = g_renew(char, irc->readbuffer, strlen(irc->readbuffer) + strlen(line) + 1);
		strcpy((irc->readbuffer + strlen(irc->readbuffer)), line);
	}

	irc_process(irc);

	/* Normally, irc_process() shouldn't call irc_free() but irc_abort(). Just in case: */
	if (!g_slist_find(irc_connection_list, irc)) {
		log_message(LOGLVL_WARNING, "Abnormal termination of connection with fd %d.", fd);
		return FALSE;
	}

	/* Very naughty, go read the RFCs! >:) */
	if (irc->readbuffer && (strlen(irc->readbuffer) > 1024)) {
		irc_abort(irc, 0, "Maximum line length exceeded");
		return FALSE;
	}

	return TRUE;
}

gboolean bitlbee_io_current_client_write(gpointer data, gint fd, b_input_condition cond)
{
	irc_t *irc = data;
	int st, size;

	size = irc->sendbuffer->len;

	if (size == 0) {
		return FALSE;
	}

	st = write(irc->fd, irc->sendbuffer->str, size);

	if (st == 0 || (st < 0 && !sockerr_again())) {
		irc_abort(irc, 1, "Write error: %s", strerror(errno));
		return FALSE;
	} else if (st < 0) { /* && sockerr_again() */
		return TRUE;
	}

	if (st == size) {
		g_string_truncate(irc->sendbuffer, 0);
		irc->w_watch_source_id = 0;

		return FALSE;
	} else {
		g_string_erase(irc->sendbuffer, 0, st);

		return TRUE;
	}
}

static gboolean bitlbee_io_new_client(gpointer data, gint fd, b_input_condition condition)
{
	socklen_t size = sizeof(struct sockaddr_in);
	struct sockaddr_in conn_info;
	int new_socket = accept(global.listen_socket, (struct sockaddr *) &conn_info, &size);

	if (new_socket == -1) {
		log_message(LOGLVL_WARNING, "Could not accept new connection: %s", strerror(errno));
		return TRUE;
	}

	if (global.conf->runmode == RUNMODE_FORKDAEMON) {
		pid_t client_pid = 0;
		int fds[2];

		if (socketpair(AF_UNIX, SOCK_STREAM, 0, fds) == -1) {
			log_message(LOGLVL_WARNING, "Could not create IPC socket for client: %s", strerror(errno));
			fds[0] = fds[1] = -1;
		}

		sock_make_nonblocking(fds[0]);
		sock_make_nonblocking(fds[1]);

		client_pid = fork();

		if (client_pid > 0 && fds[0] != -1) {
			struct bitlbee_child *child;

			/* TODO: Stuff like this belongs in ipc.c. */
			child = g_new0(struct bitlbee_child, 1);
			child->pid = client_pid;
			child->ipc_fd = fds[0];
			child->ipc_inpa = b_input_add(child->ipc_fd, B_EV_IO_READ, ipc_master_read, child);
			child->to_fd = -1;
			child_list = g_slist_append(child_list, child);

			log_message(LOGLVL_INFO, "Creating new subprocess with pid %d.", (int) client_pid);

			/* Close some things we don't need in the parent process. */
			close(new_socket);
			close(fds[1]);
		} else if (client_pid == 0) {
			irc_t *irc;

			b_main_init();

			/* Close the listening socket, we're a client. */
			close(global.listen_socket);
			b_event_remove(global.listen_watch_source_id);

			/* Make a new pipe for the shutdown signal handler */
			sighandler_shutdown_setup();

			/* Make the connection. */
			irc = irc_new(new_socket);

			/* We can store the IPC fd there now. */
			global.listen_socket = fds[1];
			global.listen_watch_source_id = b_input_add(fds[1], B_EV_IO_READ, ipc_child_read, irc);

			close(fds[0]);

			ipc_master_free_all();
		}
	} else {
		log_message(LOGLVL_INFO, "Creating new connection with fd %d.", new_socket);
		irc_new(new_socket);
	}

	return TRUE;
}

gboolean bitlbee_shutdown(gpointer data, gint fd, b_input_condition cond)
{
	/* Send the message here with now=TRUE to ensure it arrives */
	irc_write_all(TRUE, "ERROR :Closing link: BitlBee server shutting down");

	/* Try to save data for all active connections (if desired). */
	while (irc_connection_list != NULL) {
		irc_abort(irc_connection_list->data, TRUE, NULL);
	}

	/* We'll only reach this point when not running in inetd mode: */
	b_main_quit();

	return FALSE;
}
